### PR TITLE
common: add reference count helpers to RefCountedObject

### PR DESCRIPTION
**Pull request by batrick**
Main goal of this PR is to generalize the MessageInstance/MessageSubType helper classes for RefCountedObject so the technique can be applied to any derived class of RefCountedObject. Additionally, I've added opt-in protections so it's easier to prevent reference leaks, unintentionally or otherwise.

I'm tagging most leads whose code I've touched. Please review/test as needed.

### CODE REVIEW

**Comment by batrick** [comment]
I wasn't building rbd on my dev machine so I didn't catch these compiler errors. I'll fix them later this weekend...

File: src/common/RefCountedObj.h [hunk 1]
```diff
@@ -66,30 +69,128 @@ struct RefCountedObject {
...
+  RefCountedObject(CephContext* c) : cct(c) {}
+
+  virtual ~RefCountedObject() {
+    ceph_assert(nref == 0);
   }
 
-  uint64_t get_nref() const {
-    return nref;
+private:
+  mutable std::atomic<uint64_t> nref = {1};
```
Comment (liewegas) [inline src/common/RefCountedObj.h]: unrelated, but i wonder if we should make this int64_t so that we don't wrap to 0xffffffffffffffff with too many puts

**Comment by batrick** [comment]
Found this: http://tracker.ceph.com/issues/38252

Perhaps some of the RefCountedObject checks are now causing the failure to hit `make check`? ping @dillaman

File: src/common/RefCountedObj.h [hunk 2]
```diff
@@ -27,6 +27,9 @@
 
 struct RefCountedObject {
 public:
+  using ref = boost::intrusive_ptr<RefCountedObject>;
```
Comment (tchaikov) [inline src/common/RefCountedObj.h]: i'd suggest use free functions for these alias instead of embedding them as members.

**Comment by dillaman** [comment]
@batrick I asked for more details in that PR since it doesn't include any? What do you think the librbd issue is (and assuming it wasn't introduced w/ this PR)? We run all our tests under valgrind and it has never reported a memory leak.

File: src/common/RefCountedObj.h (hunk 1 above)
Comment (batrick) [reply src/common/RefCountedObj.h]: Sure, I'll change it.

File: src/common/RefCountedObj.h (hunk 2 above)
Comment (batrick) [reply src/common/RefCountedObj.h]: As an example, something like `type_ref<Message>`:

````
template<typename T>
using type_ref = boost::intrusive_ptr<T>;
````

Is this what you have in mind?

**Comment by batrick** [comment]
@dillaman it was a circular reference I introduced. Sorry about that.

`make check` is passing now. Please review :)

File: src/librbd/Journal.cc [hunk 3]
```diff
@@ -7,6 +7,7 @@
 #include "common/Timer.h"
 #include "common/WorkQueue.h"
 #include "cls/journal/cls_journal_types.h"
+#include "journal/FutureImpl.h"
```
Comment (dillaman) [inline src/librbd/Journal.cc]: Why does the private implementation header file need to be pulled in to all external client translation units?

File: src/common/RefCountedObj.h (hunk 2 above)
Comment (tchaikov) [reply src/common/RefCountedObj.h]: yes, something like that. but i don't quite like the name `type_ref`, as "type" here is redundant. probably we can have an alias in `ceph` namespace, like
```c++
namespace ceph {
template<typename T>
using ref_t = boost::intrusive_ptr<T>;
}
template<typename T>
using cref_t = boost::intrusive_ptr<T>;
}
```
as the standard library already have `std::ref<>` and `std::cref<>`.

File: src/librbd/Journal.cc (hunk 3 above)
Comment (batrick) [reply src/librbd/Journal.cc]: That was an accident, I'll remove those lines.

File: src/common/RefCountedObj.h (hunk 2 above)
Comment (batrick) [reply src/common/RefCountedObj.h]: Done, PTAL.

File: src/librbd/Journal.cc (hunk 3 above)
Comment (batrick) [reply src/librbd/Journal.cc]: Done.

**Comment by batrick** [comment]
retest this please

**Comment by batrick** [comment]
Rebased to resolve conflicts.

**Comment by batrick** [comment]
rebased

**Comment by batrick** [comment]
retest this please

**Comment by batrick** [comment]
http://pulpito.ceph.com/?branch=wip-pdonnell-testing-20190314.163335

**Comment by batrick** [comment]
retest this please

**Comment by batrick** [comment]
http://pulpito.ceph.com/?branch=wip-pdonnell-testing-20190321.172825

**Comment by batrick** [comment]
http://pulpito.ceph.com/pdonnell-2019-03-30_04:59:00-fs-wip-pdonnell-testing-20190329.211914-distro-basic-smithi/

Failures unrelated.

**Comment by batrick** [comment]
retest this please

**Comment by batrick** [comment]
retest this please

**Comment by dillaman** [comment]
http://pulpito.ceph.com/jdillaman-2019-04-02_08:42:19-rbd-wip-pdonnell-testing-20190329.211914-distro-basic-smithi/

**Comment by batrick** [comment]
rebased to resolve conflicts

**Comment by neha-ojha** [comment]
rados: http://pulpito.ceph.com/nojha-2019-04-02_01:35:01-rados-wip-refcountedobj-helpers-distro-basic-smithi/

@batrick http://pulpito.ceph.com/nojha-2019-04-02_01:35:01-rados-wip-refcountedobj-helpers-distro-basic-smithi/3799191/ is worth investigating

**Comment by batrick** [comment]
I'm going to break this into pieces to avoid the need for future rebases.

**Pull request closed without merging**